### PR TITLE
llvm10: new port

### PR DESCRIPTION
**Pull request by kencu**
###### Type(s)

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS 10.14.6 18G4032
Xcode 11.3.1 11C504 

###### Verification 
Have you

- [ ] followed our [Commit Message Guidelines](xttps://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?

- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by kencu** [comment]
This is the suggested new naming scheme, llvm-10, clang-10, lldb-10. It will require changes in various places in MacPorts and in the Portfiles to accommodate this.

If we decide to forget about this change, and stick with llvm-10.0, clang-10.0, lldb-10.0, that will probably be simpler to handle, but can be less accurate (on one occasion, the LLVM group slipped in llvm 7.1.0 due to an unexpected ABI changing bug -- but our port is still called llvm-7.0 despite it really being llvm-7.1, and this is the root issue that led to the request for a new naming scheme.

It is unlikely, but not impossible, that the LLVM group will ever again come out with an N.1.0 version of a port.

**Comment by kencu** [comment]
llvm-10 has not been tested on older systems as yet. There could be some further changes needed to allow builds on older systems.

**Comment by cjones051073** [comment]
> This is the suggested new naming scheme, llvm-10, clang-10, lldb-10. It will require changes in various places in MacPorts and in the Portfiles to accommodate this.
> 
> If we decide to forget about this change, and stick with llvm-10.0, clang-10.0, lldb-10.0, that will probably be simpler to handle, but can be less accurate (on one occasion, the LLVM group slipped in llvm 7.1.0 due to an unexpected ABI changing bug -- but our port is still called llvm-7.0 despite it really being llvm-7.1, and this is the root issue that led to the request for a new naming scheme).
> 
> It is unlikely, but not impossible, that the LLVM group will ever again come out with an N.1.0 version of a port.

FWIW, I'm in favour of using this release to drop the `.0`. Also note, even if it is returned changes are required, in PGs etc. to support the double digit major version, as the current logic assumes in a few places versions of the form `N.M` with N and M both single digits. Its true dropping the `.0` probably makes the changes required a little more complicated, but this feels to me the time to do it.

**Comment by kencu** [comment]
It would be nice to get buy-in from Marcus, Josh, and Ryan before the commit, rather than hear all about it after.

**Comment by kencu** [comment]
looks like this version of llvm no longer builds with clang-3.7, as well...so that's a bootstrapping thought for default compilers for the future...

**Comment by kencu** [comment]
hmmm ... we can't call pkg-config from the Portfile because pkgconfig is not actually installed at the time that the Portfile is parsed. 

This FFI_INCLUDES thing looks like it just never worked at all in clangs since we switched to CMAKE. -- the backticks version didn't do anything but put garbage into the cmake script, and this version can't be used with the buildbots due to the catch-22 of no pkgconfig being installed...

Perhaps if I just delete the FFI_INCLUDES setting business completely, cmake will "just find it" and go from there? Experiment coming. For now, I guess I have to cut it out of the Portfile.

File: lang/llvm-10/Portfile [hunk 1]
```diff
@@ -0,0 +1,618 @@
...
+    homepage            https://lldb.llvm.org/
+    description         the LLVM debugger
+    long_description    Lldb is the "LLVM native" debugger.
+
+    depends_lib         port:libxml2 port:libomp port:llvm-${llvm_version} port:ncurses
+    depends_build-append port:swig-python port:doxygen
+}
+
+version                 ${llvm_version}.0.0
+epoch                   1
```
Comment (cjones051073) [inline lang/llvm-10/Portfile:69]: As this is a brand new set of ports, the epoch does not need to be set, or ca be set to 0

Comment (kencu) [reply lang/llvm-10/Portfile:69]: You're right -- that's just been carried over from previous llvm's for I don't know how long...

**Comment by kencu** [comment]
> If c++14 is required now, you should perhaps set
> 
> compiler.cxx_standard 2014
> 
> this might also allow the compiler blacklisting to be simplified

Thanks for taking the time to look things over.

Because bootstrapping llvm/clang is so touchy, I/we bypass MacPorts general automatic settings and have been very specific so far as to what does and does not actually work. For this one, clang-5.0 works, as does clang-9.0 (and presumably all in between). I have not as yet figured out exactly which Xcode clang doesn't build this -- I think Xcode 8 will succeed and 7 will fail, but have to see if that holds stil..

Our bootstrapping from ancient times now looks like:

gcc-4.2 -> clang-3.4 -> clang-3.7 -> clang-9.0 -> clang-10+

with Xcode clangs joining in along the way as they can in the chain.

**Comment by cjones051073** [comment]
At the end of the day setting the c++ standard just translates into setting compiler blacklists, so in principle doing so should  not interfere with what is currently there, as it should only blacklist something already excluded... 

I think it is still worth either doing this, if only to document c++14 is required, or putting a comment into the port at the start of the blacklists explaining c++14 is now required and why the standard way of enforcing this is not done.

**Comment by kencu** [comment]
I think the mp settings work well for ports that we just want to quickly pick a usable compiler for, but it's not finely grained, and for something like this it is too coarse a setting. For example, this page <https://clang.llvm.org/cxx_status.html> confirms clang-3.4 supports all of c++14 and yet I know that it won't build this. So what I know supercedes whatever to setting says is supposed to be the case. And figguring out what people on MP have decided is a compiler that they feel deserves to be declared supporting c++14 may well skip many compilers that can build this just fine, and that can mess up bootstrapping. So that's why we don't use those settings. They are just not good enough for this.

**Comment by cjones051073** [comment]
@kencu FYI I had a little spare time so I thought I would give this MR a test.

Built it just fine on macOS 10.15

I also had a look into the changes needed to support `clang-10` as a macports compiler. In the end it was not too bad. Small changes needed both in the ports tree (to the resource files) but also to base unfortunately.

For base, see (https://github.com/macports/macports-base/pull/171). A small extension to the regex used to detect the macports clang versions was needed to allow dotless versions starting with 10. This base change can be committed independently of this MR, as it just adds support for the dotless 10+ clang versions.

For the resource changes, see this diff (https://github.com/cjones051073/macports-ports/commit/0e006d3f97f105339e76ed7a193cd45fa3210bad). A relatively minor tweak to the logic was needed, thats all.

So far in the above just as a placeholder I add `macports-clang-10` for all systems starting with macOS 10.10. That is a complete guess, so if you know how far back clang 10 goes in support, just let me know.

Finally note the changes to the `_resources` to add `macports-clang-10` to the known lists cannot be done before the MR adding support for this is in base *and* has been made available to users in a new base release. That doesn't though stop this MR being committed, as it just makes the port available.

**Comment by cjones051073** [comment]
b.t.w. I have successfully built a few ports now using `macports-clang-10` as the compiler option, using the two sets of changes above. e.g.

```
sudo port -sd install xrootd configure.compiler=macports-clang-10
```

**Comment by kencu** [comment]
Thanks, Chris! I really appreciate the extra help and testing -- messing up a core compiler is a pretty big disaster. It looks like @MarcusCalhoun-Lopez , @jmroot , and @ryandesign don't have any comments to make about the new naming scheme and what it might mean to all the Portfiles and base. I'll let this last comment go through, and if there's no reaction by the end of the day I guess it's pretty to close to being committable. I know that several of my portfiles will need to be modified to accommodate this naming scheme, should it be the final decision of everyone to go ahead with it.

**Comment by cjones051073** [comment]
> Thanks, Chris! I really appreciate the extra help and testing -- messing up a core compiler is a pretty big disaster. It looks like @MarcusCalhoun-Lopez , @jmroot , and @ryandesign don't have any comments to make about the new naming scheme and what it might mean to all the Portfiles and base. I'll let this last comment go through, and if there's no reaction by the end of the day I guess it's pretty to close to being committable. I know that several of my portfiles will need to be modified to accommodate this naming scheme, should it be the final decision of everyone to go ahead with it.

Personally I think the tweaks needed to support dropping the `.0` with version 10 are in the end quite minor really, but others should please look at my links above and make their own mind up.

Note though my point above about needing a new release of base before any reference to `macports-clang-10` can be made in any port files. This does not stop either this MR getting merged in the meantime (and I would support your going ahead with this as soon as you are happy), but until the new base support is available that is all that can be done.

**Comment by MarcusCalhoun-Lopez** [comment]
Sorry about not being responsive, but I would imagine fixing a few regular expressions in the base and PortGroups would not be too horrible.

**Comment by kencu** [comment]
OK, thank you, everyone. I guess it is time, then, to commit this. I am sure there will be some tweaking to do once we see how the mid-era systems handle this, but it built for me on 10.7, and all the newer systems. I'll likely have to tweak some clang blacklisting once we see how 10.10 to 10.12 handle it.

**Pull request merged**